satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,390 @@
+	defer mon.Task()(&ctx)(&err)
+	if message.Succeeded.GetAddressedOrderLimit() == nil {
+		return Error.New("Addressed order limit cannot be nil.")
+	}
+	if message.Succeeded.GetOriginalPieceHash() == nil {
+		return Error.New("Original piece hash cannot be nil.")
+	}
+	endpoint.log.Debug("transfer succeeded.", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+
+	// TODO validation
+
+	transfer, ok := pending.Load(message.Succeeded.OriginalPieceHash.PieceId)
+	if !ok {
+		endpoint.log.Debug("could not find transfer message in pending queue. skipping .", zap.String("piece ID", message.Succeeded.AddressedOrderLimit.Limit.PieceId.String()))
+	}
+
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	var failed int64
+	if transferQueueItem.FailedCount != nil && *transferQueueItem.FailedCount > 0 {
+		failed = -1
+	}
+
+	err = endpoint.db.IncrementProgress(ctx, nodeID, transfer.(pendingTransfer).pieceSize, 1, failed)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	pending.Delete(message.Succeeded.GetAddressedOrderLimit().GetLimit().PieceId)
+
+	return nil
+}
+
+func (endpoint *Endpoint) handleFailed(ctx context.Context, pending *sync.Map, nodeID storj.NodeID, message *pb.StorageNodeMessage_Failed) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	endpoint.log.Warn("transfer failed.", zap.String("piece ID", message.Failed.PieceId.String()), zap.String("transfer error", message.Failed.GetError().String()))
+	pieceID := message.Failed.PieceId
+	transfer, ok := pending.Load(pieceID)
+	if !ok {
+		endpoint.log.Debug("could not find transfer message in pending queue. skipping .", zap.String("piece ID", pieceID.String()))
+	}
+	transferQueueItem, err := endpoint.db.GetTransferQueueItem(ctx, nodeID, transfer.(pendingTransfer).path)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+	now := time.Now().UTC()
+	failedCount := 1
+	if transferQueueItem.FailedCount != nil {
+		failedCount = *transferQueueItem.FailedCount + 1
+	}
+
+	errorCode := int(pb.TransferFailed_Error_value[message.Failed.Error.String()])
+
+	// TODO if error code is NOT_FOUND, the node no longer has the piece. remove the queue item and the pointer
+
+	transferQueueItem.LastFailedAt = &now
+	transferQueueItem.FailedCount = &failedCount
+	transferQueueItem.LastFailedCode = &errorCode
+	err = endpoint.db.UpdateTransferQueueItem(ctx, *transferQueueItem)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// only increment failed if it hasn't failed before
+	if failedCount == 1 {
+		err = endpoint.db.IncrementProgress(ctx, nodeID, 0, 0, 1)
+		if err != nil {
+			return Error.Wrap(err)
+		}
+	}
+
+	pending.Delete(pieceID)
```

Comment:
Should we call `DeleteTransferQueueItem` if the node has exceeded the maximum failures allowed for a single piece?

Reply:
I'm still wondering about that.  We may want to have a report for failures/reasons. 

Will revisit